verification/certs.go

```diff
@@ -65,8 +170,422 @@ func getTcbInfoForHandle(c DPEClient, handle *ContextHandle) (*ContextHandle, Di
+	roots := x509.NewCertPool()
+	intermediates := x509.NewCertPool()
+
+	// Root certificate is expected to be in the beginning of the chain, the rest are expected to be intermediates.
+	roots.AddCert(certChain[0])
+
+	for _, cert := range certChain[1:] {
+		if cert.Subject.String() == cert.Issuer.String() {
+			return nil, fmt.Errorf("found a self-signed certificate in middle of certificate chain returned by GetCertificateChain")
+		}
+		intermediates.AddCert(cert)
+	}
+	opts := x509.VerifyOptions{
+		Roots:         roots,
+		Intermediates: intermediates,
+		CurrentTime:   time.Now().UTC(),
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+	}
+
+	return &opts, err
+}
+
+// Gets KeyUsage bitmap and returns as list of KeyUsage name strings.
+func getKeyUsageNames(keyUsage x509.KeyUsage) []string {
+	keyUsageNames := []string{}
+
+	if keyUsage&x509.KeyUsageDigitalSignature != 0 {
+		keyUsageNames = append(keyUsageNames, "DigitalSignature")
+	}
+
+	if keyUsage&x509.KeyUsageContentCommitment != 0 {
+		keyUsageNames = append(keyUsageNames, "ContentCommitment")
+	}
+
+	if keyUsage&x509.KeyUsageKeyEncipherment != 0 {
+		keyUsageNames = append(keyUsageNames, "KeyEncipherment")
+	}
+
+	if keyUsage&x509.KeyUsageDataEncipherment != 0 {
+		keyUsageNames = append(keyUsageNames, "DataEncipherment")
+	}
+
+	if keyUsage&x509.KeyUsageKeyAgreement != 0 {
+		keyUsageNames = append(keyUsageNames, "KeyAgreement")
+	}
+
+	if keyUsage&x509.KeyUsageCertSign != 0 {
+		keyUsageNames = append(keyUsageNames, "CertSign")
+	}
+
+	if keyUsage&x509.KeyUsageCRLSign != 0 {
+		keyUsageNames = append(keyUsageNames, "CRLSign")
+	}
+
+	if keyUsage&x509.KeyUsageEncipherOnly != 0 {
+		keyUsageNames = append(keyUsageNames, "EncipherOnly")
+	}
+
+	if keyUsage&x509.KeyUsageDecipherOnly != 0 {
+		keyUsageNames = append(keyUsageNames, "DecipherOnly")
+	}
+
+	return keyUsageNames
+}
+
+// Verifies the TCI_Current and TCI_Cumulative of dice tcb information blocks
+func verifyDiceTcbDigest(tcbInfo DiceTcbInfo, wantCurrentTCI []byte, lastCumulativeTCI []byte) error {
+	var err error
+
+	// Check TCI_CURRENT
+	currentTCI := tcbInfo.Fwids[0].Digest
+	if !bytes.Equal(currentTCI, wantCurrentTCI) {
+		err = fmt.Errorf("unexpected TCI_CURRENT digest, want %v but got %v", wantCurrentTCI, currentTCI)
+	}
+
+	// Check TCI_CUMULATIVE against expected cumulative TCI
+	wantCumulativeTCI := computeExpectedCumulative(lastCumulativeTCI, currentTCI)
+	cumulativeTCI := tcbInfo.Fwids[1].Digest
+	if !bytes.Equal(cumulativeTCI, wantCumulativeTCI) {
+		err = fmt.Errorf("unexpected TCI_CUMULATIVE value, want %v but got %v", wantCumulativeTCI, cumulativeTCI)
+	}
+	return err
+}
+
+// Checks the FWID block's Digest.
+// FWID at index 0 has the TCI_CURRENT as digest
+// FWID at index 1 has the TCI_CUMULATIVE as digest
+// The length of FWID array in each DICE TCB information block is 2.
+func validateDiceTcbFwids(leafCertBytes []byte, expectedCurrentTci []byte) error {
+	var leafCert *x509.Certificate
```

Comment:
This function cycles through all multi TCB info blocks and validates cumulative measurement